{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Fast blocked Gibbs sampling\n",
        "\n",
        "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/QAHRoddur/JWAS/blob/main/Examples/fast_blocked_gibbs_sampling.ipynb)\n",
        "\n",
        "This notebook is auto-generated from the JWAS.jl wiki page.\n"
      ],
      "id": "1afa5128de6a",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "using Pkg\n",
        "Pkg.add(\"JWAS\")\n",
        "Pkg.precompile()\n",
        "using JWAS\n"
      ],
      "id": "a1a13346ef68",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "#  Fast blocked Gibbs sampling\n",
        "The fast blocked Gibbs sampling for Bayesian linear mixed model is proposed in below papers: \n",
        "> Breen, E.J., MacLeod, I.M., Ho, P.N. et al. BayesR3 enables fast MCMC blocked processing for largescale multi-trait genomic prediction and QTN mapping analysis. Commun Biol 5, 661 (2022). https://doi.org/10.1038/s42003-022-03624-1\n",
        "\n",
        "> Zhao, T., Fernando, R., Garrick, D. et al. Fast parallelized sampling of Bayesian regression models for whole-genome prediction. Genet Sel Evol 52, 16 (2020). https://doi.org/10.1186/s12711-020-00533-x\n",
        "\n",
        "\n",
        "\n",
        "# JWAS\n",
        "The fast block mode can be activated via the `fast_blocks` argument in the `runMCMC()` function. The `fast_blocks` argument can be assigned the number of SNPs in each block (i.e., block size). For example, for 50k SNP, when `chain_length=1000` and `fast_blocks=100`, there will be 500 blocks, each containing 100 SNPs (i.e., block size=100 SNPs). If `fast_blocks=true`, then block size is the square root of the number of individuals. Note that when the fast_blocks mode is enabled, the number of SNPs in each block will be displayed in JWAS printouts, e.g., \"BLOCK SIZE: 100\".\n",
        "\n",
        "**JWAS code to determine block size**:\n",
        "```{julia}\n",
        "if fast_blocks == true\n",
        "    block_size = Int(floor(sqrt(mme.M[1].nObs)))\n",
        "elseif typeof(fast_blocks) <: Number\n",
        "    block_size = Int(floor(fast_blocks))\n",
        "end\n"
      ],
      "id": "53fe39765761",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "**JWAS structure for outer and inner iterations**:\n",
        "\n",
        "#total iterations = #outer iterations $*$ #inner iterations\n",
        "\n",
        "```{julia}\n",
        "#MCMC iteration start\n",
        "for k in 1:outer_niter\n",
        "    for i in 1:blocks\n",
        "        nreps = block_size #nreps=block_size, outer_niter=niter/block_size\n",
        "        for reps = 1:nreps\n",
        "            for j=1:block_size\n",
        "                #sample marker effects\n",
        "\t    end\n",
        "\tend\n",
        "    end \n",
        "end\n",
        "#MCMC iteration end\n"
      ],
      "id": "dc547fdddc2d",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "*  for example: 50k SNPs, block size = 100 SNPs (thus 500 blocks), 2000 total iterations\n",
        "* 2000 total iterations = 20  outer iterations  * 100 inner iterations\n",
        "* in each outer iteration:  collect 100 MCMC samples for 100 SNPs in block1,......, collect 100 MCMC samples for 100 SNPs in block500 --> in total, we collect 100 MCMC samples for 50k SNPs.\n",
        "* in each outer iteration, we have 100 MCMC samples for all SNPs. --> Thus, for all 20 outer iterations, we will have 2000 MCMC samples for all SNPs.\n",
        "\n",
        "\n",
        "# JWAS examples\n",
        "Note that `fast_blocks` now only works for one class of genotype, so `fast_blocks` cannot be used for [Multi-class Bayesian Analysis](https://github.com/reworkhow/JWAS.jl/wiki/Multiclass-Bayesian-analysis).\n",
        "\n",
        "## Univariate Linear Mixed Model (`fast_blocks`)\n",
        "```{julia}\n",
        "# Step 1: Load packages\n",
        "using JWAS,DataFrames,CSV,Statistics,JWAS.Datasets, Random\n",
        "Random.seed!(123)\n",
        "\n",
        "# Step 2: Read data\n",
        "phenofile = dataset(\"phenotypes.txt\", dataset_name=\"demo_7animals\")\n",
        "pedfile = dataset(\"pedigree.txt\", dataset_name=\"demo_7animals\")\n",
        "genofile = dataset(\"genotypes.txt\", dataset_name=\"demo_7animals\")\n",
        "phenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstring=[\"NA\"])\n",
        "pedigree = get_pedigree(pedfile,separator=\",\",header=true);\n",
        "genotypes = get_genotypes(genofile,separator=',',method=\"BayesC\");\n",
        "first(phenotypes,5)\n",
        "\n",
        "# Step 3: Build Model Equations\n",
        "model_equation =\"y1 = intercept + x1 + x2 + x2*x3 + ID + dam + genotypes\";\n",
        "model = build_model(model_equation);\n",
        "\n",
        "# Step 4: Set Factors or Covariates\n",
        "set_covariate(model,\"x1\");\n",
        "\n",
        "# Step 5: Set Random or Fixed Effects\n",
        "set_random(model,\"x2\");\n",
        "set_random(model,\"ID dam\",pedigree);\n",
        "\n",
        "# Step 6: Run Analysis\n",
        "out=runMCMC(model,phenotypes,chain_length = 1000, fast_blocks = 100);\n",
        "\n",
        "# Step 7: Check Results\n",
        "out[\"EBV_y1\"]  # estimated breeding values\n"
      ],
      "id": "585187d33fed",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "## Single-trait Single-step Analysis (`fast_blocks`)\n",
        "```{julia}\n",
        "# Step 1: Load packages\n",
        "using JWAS,DataFrames,CSV,Statistics,JWAS.Datasets, Random\n",
        "Random.seed!(123)\n",
        "\n",
        "# Step 2: Read data\n",
        "phenofile = \"../data/phenotypes.txt\"\n",
        "pedfile = \"../data/pedigree.txt\"\n",
        "genofile = \"../data/genotypes.txt\"\n",
        "phenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstring=[\"NA\"])\n",
        "pedigree = get_pedigree(pedfile,separator=\",\",header=true);\n",
        "genotypes = get_genotypes(genofile,separator=',',method=\"BayesC\");\n",
        "first(phenotypes,5)\n",
        "\n",
        "# Step 3: Build Model Equations\n",
        "model_equation =\"y1 = intercept + x1 + x2 + x2*x3 + ID + dam + genotypes\";\n",
        "model = build_model(model_equation);\n",
        "\n",
        "# Step 4: Set Factors or Covariates\n",
        "set_covariate(model,\"x1\");\n",
        "\n",
        "# Step 5: Set Random or Fixed Effects\n",
        "set_random(model,\"x2\");\n",
        "set_random(model,\"ID dam\",pedigree);\n",
        "\n",
        "# Step 6: Run Analysis\n",
        "out=runMCMC(model,phenotypes,single_step_analysis=true,pedigree=pedigree,chain_length = 1000, fast_blocks = 100);\n",
        "\n",
        "# Step 7: Check Results\n",
        "out[\"EBV_y1\"]  # estimated breeding values\n"
      ],
      "id": "2e7631284536",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Multivariate Linear Mixed Model (`fast_blocks`)\n",
        "```{julia}\n",
        "# Step 1: Load packages\n",
        "using JWAS,DataFrames,CSV,Statistics,JWAS.Datasets, Random\n",
        "Random.seed!(123)\n",
        "\n",
        "# Step 2: Read data\n",
        "phenofile =  dataset(\"phenotypes.txt\", dataset_name=\"demo_7animals\")\n",
        "pedfile =  dataset(\"pedigree.txt\", dataset_name=\"demo_7animals\")\n",
        "genofile =  dataset(\"genotypes.txt\", dataset_name=\"demo_7animals\")\n",
        "phenotypes = CSV.read(phenofile,DataFrame,delim =  ',',header=true,missingstring=[\"NA\"])\n",
        "pedigree =  get_pedigree(pedfile,separator=\",\",header=true);\n",
        "genotypes =  get_genotypes(genofile,separator=',',method=\"BayesC\");\n",
        "first(phenotypes,5)\n",
        "\n",
        "# Step 3: Build Model Equations\n",
        "model_equation =\"y1 = intercept + x1 + x2 + x2*x3 + ID + dam + genotypes\n",
        "                 y2 = intercept + x1 + x2 + ID + genotypes\n",
        "                 y3 = intercept + x1 + ID + genotypes\";\n",
        "model =  build_model(model_equation);\n",
        "\n",
        "# Step 4: Set Factors or Covariates\n",
        "set_covariate(model,\"x1\");\n",
        "\n",
        "# Step 5: Set Random or Fixed Effects\n",
        "set_random(model,\"x2\");\n",
        "set_random(model,\"ID dam\",pedigree);\n",
        "\n",
        "# Step 6: Run Analysis\n",
        "out=runMCMC(model,phenotypes,chain_length =  1000, fast_blocks =  100);\n",
        "\n",
        "# Step 7: Check Results\n",
        "out[\"EBV_y3\"]  # estimated breeding values\n"
      ],
      "id": "25d36b7f79a8",
      "metadata": {}
    }
  ],
  "nbformat_minor": 5,
  "metadata": {
    "colab": {
      "provenance": [],
      "name": "Fast blocked Gibbs sampling.ipynb"
    },
    "language_info": {
      "file_extension": ".jl",
      "name": "julia"
    },
    "kernelspec": {
      "name": "julia-1.11",
      "display_name": "Julia 1.11",
      "language": "julia"
    }
  },
  "nbformat": 4
}