{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Ramdom Regression Marker Effect Model\n",
        "\n",
        "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/QAHRoddur/JWAS/blob/main/Examples/ramdom_regression_marker_effect_model.ipynb)\n",
        "\n",
        "This notebook is auto-generated from the JWAS.jl wiki page.\n"
      ],
      "id": "6f583840b0c6",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "using Pkg\n",
        "Pkg.add(\"JWAS\")\n",
        "Pkg.precompile()\n",
        "using JWAS\n"
      ],
      "id": "f44b302dff84",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Example of Random regression model with mixture prior\n",
        "\n",
        "paper: \n",
        "> * Qu, Jiayi, Gota Morota, and Hao Cheng. \"A Bayesian random regression method using mixture priors for genome‐enabled analysis of time‐series high‐throughput phenotyping data.\" The Plant Genome 15, no. 3 (2022): e20228; https://doi.org/10.1002/tpg2.20228.\n"
      ],
      "id": "0b4c24ff3811",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "# Load packages\n",
        "using JWAS, DataFrames, CSV, DelimitedFiles, LinearAlgebra, Statistics, Random;\n",
        "using GSL,Kronecker;\n",
        "using Plots;\n",
        "using HTTP;\n",
        "\n",
        "# read data from github\n",
        "function dataset_big(file_name::AbstractString; dataset_url::AbstractString=\"https://raw.githubusercontent.com/zhaotianjing/bio_protocol/refs/heads/main/data/\")\n",
        "\n",
        "    rdaname = joinpath(dataset_url, string(file_name))\n",
        "    http_obj = HTTP.get(rdaname)\n",
        "    df = CSV.File(IOBuffer(http_obj.body)) |> DataFrame\n",
        "    #df = CSV.File(IOBuffer(http_obj.body); delim = ',', missingstring=\"NA\", types=Dict(:ID => String)) |> DataFrame \n",
        "    return df\n",
        "end\n",
        "\n",
        "phenotypes = dataset_big(\"RR_PSA_rep1.csv\")\n",
        "genotypes = dataset_big(\"Genotype2_first5k.csv\")\n",
        "# simulated true QTL\n",
        "trueQTLs = dataset_big(\"QTLpos_rep1.csv\")\n"
      ],
      "id": "eb6cf1d60776",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "In this code, we assume that **Legendre polynomials** are used to model the standardized time points. The **order of the polynomials** is determined by the parameter `ncoef`, which specifies the number of polynomial coefficients to be estimated. The function `generatefullPhi(timevec, ncoeff)` constructs the **Legendre polynomial matrix (Φ)** for a given set of time points. Each row of the matrix represents a time point, while each column corresponds to a Legendre polynomial of increasing order.\n",
        " \n"
      ],
      "id": "e6508364cb7e",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "ncoef = 2\n",
        "### Generate Lengendre Polynomial Covariates\n",
        "function generatefullPhi(timevec, ncoeff=3)\n",
        "    # timevec: a vector of time points for all observations\n",
        "    # ncoeff : the number of polynomial coefficients to be estimated\n",
        "    times = sort(unique(timevec)) # vector of whole timepoints sorted\n",
        "    tmin = minimum(times)\n",
        "    tmax = maximum(times)\n",
        "    # standardized time points\n",
        "    qi = 2 * (times .- tmin) ./ (tmax .- tmin) .- 1\n",
        "    Φ = Matrix{Float64}(undef, length(times), ncoeff) # Phi of size ntimes x ncoeff\n",
        "    for i = 1:ncoeff\n",
        "        n = i - 1\n",
        "        ## Given the normalized function from L. R. Schaeffer\n",
        "        Φ[:, i] = sqrt((2 * n + 1) / 2) * sf_legendre_Pl.(n, qi) # construct the matrix\n",
        "    end\n",
        "    return Φ\n",
        "end\n",
        "# the Ledengre polynomial matrix \n",
        "myPhi = generatefullPhi(phenotypes[!, :time], ncoef)\n"
      ],
      "id": "899b52511477",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "This part of the code splits the dataset into training and testing subsets for genomic prediction.\n"
      ],
      "id": "e02da29007cc",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "# number of testing individual = 119 (1/3)\n",
        "IDs = unique(phenotypes[!, :ID])\n",
        "nind = size(IDs, 1)\n",
        "nt = 119\n",
        "testingInd = shuffle(IDs)[1:nt]\n",
        "trainingInd = filter!(x -> x ∉ testingInd, IDs);\n",
        "pheno_training = filter(row -> row.ID in trainingInd, copy(phenotypes))\n",
        "pheno_testing = filter(row -> row.ID in testingInd, copy(phenotypes));\n",
        "\n",
        "# Add Legendre Polynomial Covariates to Training Data\n",
        "wholePhi = myPhi[pheno_training[!, :time], :]\n",
        "\n",
        "for i in 1:ncoef\n",
        "    pheno_training[!, Symbol(\"Phi$i\")] = wholePhi[:, i]\n",
        "end\n",
        "\n",
        "# Add Day$i columns to pheno_training to account for heterogeneous residual variance\n",
        "tp = length(unique(pheno_training[!, :time]))\n",
        "\n",
        "for i in 1:tp\n",
        "    colname = Symbol(\"Day$i\")\n",
        "    colvalue = Vector{Union{Missing,Int}}(missing, size(pheno_training, 1))  # Allows missing + int\n",
        "    index = findall(pheno_training[!, :time] .== i)\n",
        "    colvalue[index] .= 1:length(index)  # Assign sequential values\n",
        "    pheno_training[!, colname] = colvalue\n",
        "end\n",
        "\n",
        "# Add a column to pheno_training to represent permanent environmental effects\n",
        "pheno_training[!, :pe] = pheno_training[!, :ID]\n"
      ],
      "id": "6c18144c699e",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Fit the RR-BayesC model\n"
      ],
      "id": "16055c073e56",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "# get_genotypes\n",
        "geno = get_genotypes(genotypes, header=true, separator=',',method=\"BayesC\")\n",
        "\n",
        "#model equation \n",
        "fixed_effect = join(\" + Phi\" .* string.(collect(2:ncoef)))\n",
        "\n",
        "# add permanent environmental effects \n",
        "pe = join(\" + Phi\" .* string.(collect(1:ncoef)) .* \"*pe\")\n",
        "\n",
        "#add heterogeneous variance \n",
        "hetero_residual = join(\" + Day\" .* string.(collect(1:tp)))\n",
        "\n",
        "model_equation = \"PSA = Phi1\" * fixed_effect * pe * hetero_residual * \"+ geno\"\n",
        "\n",
        "model = build_model(model_equation);\n",
        "\n",
        "#Set Model Parameters\n",
        "\n",
        "# set covariates\n",
        "set_covariate(model, \"Phi1\")\n",
        "set_covariate(model, \"Phi2\");\n",
        "\n",
        "# set random effects for pe\n",
        "set_random(model, \"Phi1*pe\")\n",
        "set_random(model, \"Phi2*pe\")\n",
        "\n",
        "# set random effects for heterogeneous residual variance\n",
        "for d in 1:tp\n",
        "    set_random(model, \"Day$d\")\n",
        "end\n",
        "\n",
        "# run MCMC sampling \n",
        "@time output = runMCMC(model, pheno_training, burnin=10,chain_length=100, RRM=myPhi, seed=123);\n"
      ],
      "id": "68801b50ebbd",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "This section of the code estimates time-specific breeding values (BVs) for individuals by using estimated breeding values (EBVs) of random regression coefficients. It also evaluates prediction accuracy for testing individuals across different time points.\n"
      ],
      "id": "f9db39819243",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "#### Validation \n",
        "# estimated EBVs (coefficients) for individuals\n",
        "u1 = CSV.read(\"results/EBV_1.txt\", DataFrame)\n",
        "u2 = CSV.read(\"results/EBV_2.txt\", DataFrame)\n",
        "EBVs = DataFrame(ID=u1[!, :ID],u1=u1[!, :EBV],u2=u2[!, :EBV])\n",
        "\n",
        "# Compute gBLUP for Each Time Point\n",
        "# Calculate the BV for individuals by time \n",
        "n = size(EBVs, 1)\n",
        "Idt = Matrix{Float64}(I, n, n)\n",
        "Phi = generatefullPhi(collect(1:tp), ncoef)\n",
        "Z = Idt ⊗ Phi\n",
        "us = [[] for i = 1:n]\n",
        "for i = 1:n\n",
        "    us[i] = Matrix(EBVs)[i, 2:end]\n",
        "end\n",
        "uhat = vcat(us...)\n",
        "gBLUP = Z * uhat\n",
        "gBLUP = convert(Array{Float64,1}, gBLUP)\n",
        "\n",
        "#Organize time-specific EBVs by time\n",
        "correctEBVs = DataFrame(\n",
        "    ID=repeat(EBVs[!, :ID], inner=tp),\n",
        "    time=repeat(collect(1:tp), outer=n),\n",
        "    EBV=gBLUP)\n",
        "\n",
        "# choose EBV for testing Individuals\n",
        "testingEBVs = filter(row -> row.ID in testingInd, correctEBVs)\n",
        "\n",
        "# calculate prediction accuracy across time for testing individuals\n",
        "accuracy = []\n",
        "for timei = 1:tp\n",
        "    println(\"Day$timei.\")\n",
        "    pheno_testingi = filter(row -> row.time in [timei], pheno_testing)\n",
        "    select!(pheno_testingi, Not(:time))\n",
        "    EBV_testingi = filter(row -> row.time in [timei], testingEBVs)\n",
        "    select!(EBV_testingi, Not(:time))\n",
        "    final_dfi = innerjoin(pheno_testingi, EBV_testingi, on=:ID)\n",
        "    res = cor(final_dfi[!, :EBV], final_dfi[!, :PSA])\n",
        "    push!(accuracy, res)\n",
        "end\n"
      ],
      "id": "aa040b2336cf",
      "metadata": {},
      "execution_count": null
    },
    {
      "cell_type": "markdown",
      "source": [
        "This section of the code estimates the posterior inclusion probability (PIP) for single nucleotide polymorphisms (SNPs) that influence a specific random regression coefficient in a Bayesian regression model.\n"
      ],
      "id": "3381fabac9bf",
      "metadata": {}
    },
    {
      "outputs": [],
      "cell_type": "code",
      "source": [
        "# calculate the model frequency for coef i (1: intercept, 2: slope)\n",
        "coef = 2\n",
        "println(\"Coef $coef\")\n",
        "mrk_eff_sample_c = \"results/MCMC_samples_marker_effects_geno_$coef.txt\"\n",
        "mrk_samples, markerID = readdlm(mrk_eff_sample_c, header=true, ',')\n",
        "markerID = vec(strip.(markerID, ['\\\"']))\n",
        "model_freq = vec(mean(mrk_samples .!= 0, dims=1))\n",
        "# markerOrder is used to order SNPs in the plots\n",
        "RR_MFc = DataFrame(markerID=markerID, modelfrequency=model_freq)\n"
      ],
      "id": "3cad9b5a41f0",
      "metadata": {},
      "execution_count": null
    }
  ],
  "nbformat_minor": 5,
  "metadata": {
    "colab": {
      "provenance": [],
      "name": "Ramdom Regression Marker Effect Model.ipynb"
    },
    "language_info": {
      "file_extension": ".jl",
      "name": "julia"
    },
    "kernelspec": {
      "name": "julia-1.11",
      "display_name": "Julia 1.11",
      "language": "julia"
    }
  },
  "nbformat": 4
}